crates/ntx-builder/src/builder/server/state.rs

```diff
@@ -105,10 +105,10 @@ impl PendingNotes {
                     .expect("note must be on the map if nullifier was found")
                     .metadata()
                     .tag();
-
-                let bucket =
-                    self.by_tag.get_mut(&tag).expect("any tracked note is in the by_tag map");
-                bucket.retain(|&x| x != id);
+                // The bucket may have been pruned already, so check before
+                if let Some(bucket) = self.by_tag.get_mut(&tag) {
+                    bucket.retain(|&x| x != id);
+                }
```

Comment:
Fix 3: As the comment mentions, the bucket may be pruned at this point so it's better to have a sanity check